lib/docker_tag_pusher.rb

```diff
@@ -0,0 +1,73 @@
+require 'json'
+require 'net/http'
+require 'uri'
+
+class DockerTagPusher
+  REGISTRY = 'https://registry-1.docker.io'.freeze
+  MEDIA_TYPE = 'application/vnd.docker.distribution.manifest.v2+json'.freeze
+
+  attr_reader :username, :password
+  private :username, :password
+
+  def initialize(username, password)
+    @username = username
+    @password = password
+  end
+
+  def get_manifest(repo, tag)
+    request = Net::HTTP::Get.new(
+      "#{REGISTRY}/v2/#{repo}/manifests/#{tag}",
+      'Authorization' => "Bearer #{token(repo)}",
+      'Accept' => MEDIA_TYPE
+    )
+    response = registry_client.request(request)
+
+    if response.code == '404'
+      raise 'Image or tag not found'
+    elsif response['Content-Type'] != MEDIA_TYPE
+      raise "Remote image not in correct format (must be #{MEDIA_TYPE})"
+    elsif response.code != '200'
```

Comment:
Will `Net::HTTP::Get` above follow redirects? If it doesn't by default, then is that something we want it to do or is it reasonable to error instead? I don't know if this will ever happen because I've no experience with docker registry and such.

Reply:
I don't think it's possible for the registry to return a redirect for the endpoints we're calling here.